flooder/src/main.rs

```diff
@@ -36,50 +38,89 @@ fn main() -> Result<(), anyhow::Error> {
     }
-
     let rate_limiting = match (config.transactions_in_interval, config.interval_secs) {
         (Some(tii), Some(is)) => Some((tii, is)),
         (None, None) => None,
         _ => panic!("--transactions-in-interval needs to be specified with --interval-secs"),
     };
-
-    // we want to fail fast in case seed or phrase are incorrect
-    if !config.skip_initialization && config.phrase.is_none() && config.seed.is_none() {
-        panic!("Needs --phrase or --seed");
-    }
-    let pool = create_connection_pool(&config.nodes);
-    let connection = pool.get(0).unwrap();
-    let tx_status = match config.submit_only {
-        true => XtStatus::SubmitOnly,
-        false => XtStatus::Ready,
+    let tx_status = match config.wait_for_ready {
+        true => XtStatus::Ready,
+        false => XtStatus::SubmitOnly,
     };
+    info!(
+        "initializing thread-pool: {}ms",
+        time_stats.elapsed().as_millis()
+    );
+    let mut thread_pool_builder = rayon::ThreadPoolBuilder::new();
+    if let Some(threads) = config.threads {
+        let threads = threads.try_into().expect("`threads` within usize range");
+        thread_pool_builder = thread_pool_builder.num_threads(threads);
+    }
+    let thread_pool = thread_pool_builder
+        .build()
+        .expect("thread-pool should be created");
 
     info!(
-        "Preparing transactions: {}ms",
+        "thread-pool initialized: {}ms",
+        time_stats.elapsed().as_millis()
+    );
+    let threads = thread_pool.current_num_threads();
+    info!("thread-pool reported {} threads", threads);
+    // each thread should have its own connection
+    info!(
+        "creating connection-pool: {}ms",
         time_stats.elapsed().as_millis()
     );
-    let txs = preparing_txs(&config, connection);
+    let pool = create_connection_pool(&config.nodes, threads);
     info!(
-        "Transactions prepared: {}ms",
+        "connection-pool created: {}ms",
+        time_stats.elapsed().as_millis()
+    );
+    let connection = pool.get(0).unwrap().clone();
+
+    info!(
+        "preparing transactions: {}ms",
+        time_stats.elapsed().as_millis()
+    );
+    let txs = thread_pool.install(|| prepare_txs(&config, &connection));
+    info!(
+        "transactions prepared: {}ms",
         time_stats.elapsed().as_millis()
     );
 
     let histogram = Arc::new(Mutex::new(
         HdrHistogram::<u64>::new_with_bounds(1, u64::MAX, 3).unwrap(),
     ));
 
+    let (transactions_in_interval, interval_duration) = rate_limiting.map_or(
+        (txs.len(), Duration::from_secs(0)),
+        |(transactions_in_interval, secs)| {
+            (
+                transactions_in_interval
+                    .try_into()
+                    .expect("`transactions_in_interval` should be within usize range"),
+                Duration::from_secs(secs),
+            )
+        },
+    );
+
+    let pool_getter = |tx_ix: usize| pool.get(tx_ix % pool.len()).unwrap();
```

Comment:
Correct me if I am wrong but this means that different threads can get the same connection from the pool. We should have a distinct connection for each node for each thread to leverage the parallelism.

Reply:
Please see the new version.